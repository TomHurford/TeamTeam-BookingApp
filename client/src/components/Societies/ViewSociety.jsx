import React, { useEffect, useState } from "react";
import Event from "../Events/Event";
import { useParams } from "react-router-dom";
import axios from "axios";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import {
  faFacebook,
  faInstagram,
  faTwitter,
} from "@fortawesome/free-brands-svg-icons";
import ContactSocietyForm from "./ContactSocietyForm";
import { Link } from "react-router-dom";
import "../../styles/index.css";
const jwtController = require("../../utils/jwt.js");

function ViewSociety() {
  const [society, setSociety] = useState({});
  const [societyLinks, setSocietyLinks] = useState({});
  const [showFollowButton, setShowFollowButton] = useState(true);
  const [showUnfollowButton, setShowUnfollowButton] = useState(true);
  const [showEditButton, setShowEditButton] = useState(true);
  const [events, setEvents] = useState([]);
  const { id: societyId } = useParams();
  const data = {
    societyId: parseInt(societyId),
  };

  useEffect(() => {
    if (
      jwtController.getToken() === undefined ||
      jwtController.getToken() === null
    ) {
      axios
        .post("http://localhost:5001/societies/getSociety", {
          societyId: societyId,
        })
        .then((response) => {
          setSociety(response.data.society);
          setSocietyLinks(response.data.society.links[0]);
          setEvents(response.data.society.events);
        })
        .catch((error) => {
          console.log(error);
        });
    } else {
      fetch("http://localhost:5001/societies/getSociety", {
        method: "POST",
        mode: "cors",
        headers: {
          "Content-Type": "application/json",
          Authorization: "Bearer " + jwtController.getToken(),
        },
        body: JSON.stringify({ societyId: parseInt(societyId) }),
      })
        .then((response) => response.json())
        .then((data) => {
          setSociety(data.society);
          setSocietyLinks(data.society.links[0]);
          setEvents(data.society.events);
        })
        .catch((error) => {
          console.log(error);
        });
    }
  }, [societyId]);

  useEffect(() => {
    if (
      jwtController.getToken() !== undefined &&
      jwtController.getToken() !== null
    ) {
      fetch("http://localhost:5001/societies/checkUserIsMember", {
        method: "POST",
        headers: {
          "Content-Type": "application/json",

          Authorization: "Bearer " + jwtController.getToken(),
        },
        body: JSON.stringify({ societyId: parseInt(societyId) }),
      }).then((res) => {
        res.json().then((data) => {
          if (data.isMember === true) {
            setShowFollowButton(false);
            setShowUnfollowButton(true);
          } else {
            setShowFollowButton(true);
            setShowUnfollowButton(false);
          }
        });
      });
    }
  }, [societyId]);

  // useEffect(() => {
  //   fetch("http://localhost:5001/societies/getFollowedSocieties", {
  //     method: "POST",
  //     headers: {
  //       "Content-Type": "application/json",
  //       Authorization: "Bearer " + jwtController.getToken(),
  //     },
  //     body: JSON.stringify(data),
  //   }).then((res) => {
  //     res.json().then((data) => {
  //       for (let i = 0; i < data.societies.length; i++) {
  //         if (data.societies[i].societyId === parseInt(societyId)) {
  //           setShowFollowButton(false);
  //         }
  //       }
  //     });
  //   });
  // }, [societyId]);

  useEffect(() => {
    if (
      jwtController.getToken() !== undefined &&
      jwtController.getToken() !== null
    ) {
      console.log();
      fetch("http://localhost:5001/societies/checkPresident", {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          "Authorization": "Bearer " + jwtController.getToken(),
        },
        body: JSON.stringify(data),
      }).then((res) => {
        res.json().then((data) => {
          if (data.isPresident === false) {
            setShowEditButton(false);
          }
        });
      });
    }
  }, [societyId]);

  function societyEventClick(eventId) {
    window.location.href = "/event-details?eventId=" + eventId;
  }
  //Render the events for the particular society
  function eventsCardList(events) {
    return (
      <div className="events" data-testid="events-list">
        {events.map((event) => (
          <div
            data-testid={"event" + event.id}
            className="eventCard"
            key={event.id}
            onClick={() => societyEventClick(event.id)}
          >
            <Event details={event.id} specificEvent={event} />
          </div>
        ))}
      </div>
    );
  }

  function followSociety() {
    fetch("http://localhost:5001/societies/followSociety", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: "Bearer " + jwtController.getToken(),
      },
      body: JSON.stringify(data),
    })
      .then((res) => res.status)
      .then((status) => {
        if (status === 200) {
          setShowFollowButton(false);
          setShowUnfollowButton(true);
          alert("Society followed successfully!");
        } else {
          alert("Error following society");
        }
      });
  }

  function unfollowSociety() {
    fetch("http://localhost:5001/societies/unFollowSociety", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: "Bearer " + jwtController.getToken(),
      },
      body: JSON.stringify(data),
    })
      .then((res) => res.status)
      .then((status) => {
        if (status === 200) {
          setShowUnfollowButton(false);
          setShowFollowButton(true);
          alert("Society unfollowed successfully!");
        } else {
          alert("Error unfollowing society");
        }
      });
  }

  return (
    <div className="page-container" id="societyPage">
      <div className="underlay"></div>

      <div className="header">
        <div
          className="image"
          style={{ backgroundImage: `url(${societyLinks.banner})` }}
        >
          <div className="bannerOverlay"></div>
        </div>
        <div className="identity">
          <div className="icon">
            <div
              className="logo"
              style={{ backgroundImage: `url(${societyLinks.logo})` }}
            ></div>
          </div>
          <div className="name">{society.name}</div>
          <div className="socials" style={{ marginRight: "10px" }}>
            <a href={societyLinks.facebook} className="socialCircle">
              <img
                src="https://upload.wikimedia.org/wikipedia/commons/thumb/0/05/Facebook_Logo_%282019%29.png/1024px-Facebook_Logo_%282019%29.png"
                alt="facebook"
              />
            </a>
            <a href={societyLinks.twitter} className="socialCircle">
              <img
                src="https://upload.wikimedia.org/wikipedia/commons/thumb/4/4f/Twitter-logo.svg/2491px-Twitter-logo.svg.png"
                alt="twitter"
              />
            </a>
            <a href={societyLinks.instagram} className="socialCircle">
              <img
                src="https://upload.wikimedia.org/wikipedia/commons/thumb/e/e7/Instagram_logo_2016.svg/2048px-Instagram_logo_2016.svg.png"
                alt="instagram"
              />
            </a>
          </div>
        </div>
      </div>

      <div className="body">
        <div className="description">
          <p>{society.description}</p>
          <p>
            <strong>Category:</strong> {society.category}
          </p>
          <p>
            <strong>Followers:</strong> {society.members}
          </p>
          <a href={societyLinks.website}>{societyLinks.website}</a>
          <div>
            {jwtController.getToken() !== undefined &&
              jwtController.getToken() !== null &&
              showFollowButton && (
                <button
                  data-testid="followButton"
                  type="button"
                  className="button"
                  onClick={followSociety}
                >
                  Follow
                </button>
              )}

            {jwtController.getToken() !== undefined &&
              jwtController.getToken() !== null &&
              showUnfollowButton && (
                <button
                  type="button"
                  data-testid="unfollowButton"
                  className="button button--red"
                  onClick={unfollowSociety}
                >
                  Unfollow
                </button>
              )}
          </div>
        </div>

        <div className="events">
          <h2>Events:</h2>
          {eventsCardList(events)}
        </div>
      </div>

      <ContactSocietyForm
        societyName={society.name}
        societyEmail={society.email}
      />

      <Link to={`/edit-society/${society.id}`}>
        {society.isCommitteePresident === true && showEditButton && (
          <button type="button" className="button">
            Edit Society
          </button>
        )}
      </Link>
    </div>
  );
}

export default ViewSociety;
