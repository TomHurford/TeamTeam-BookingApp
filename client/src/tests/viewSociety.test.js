import React from 'react';
import renderer from 'react-test-renderer';
import ViewSociety from '../components/Societies/ViewSociety';

test('ViewSociety snapshot test', () => {
  const component = renderer.create(<ViewSociety />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});