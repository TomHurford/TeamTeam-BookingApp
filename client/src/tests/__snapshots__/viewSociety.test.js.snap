// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewSociety snapshot test 1`] = `
<div
  className="page-container"
  id="societyPage"
>
  <div
    className="underlay"
  />
  <div
    className="header"
  >
    <div
      className="image"
      style={
        Object {
          "backgroundImage": "url(undefined)",
        }
      }
    >
      <div
        className="bannerOverlay"
      />
    </div>
    <div
      className="identity"
    >
      <div
        className="icon"
      >
        <div
          className="logo"
          style={
            Object {
              "backgroundImage": "url(undefined)",
            }
          }
        />
      </div>
      <div
        className="name"
      />
      <div
        className="socials"
      >
        <a
          className="socialCircle"
        >
          <img
            alt="facebook"
            src="https://upload.wikimedia.org/wikipedia/commons/thumb/0/05/Facebook_Logo_%282019%29.png/1024px-Facebook_Logo_%282019%29.png"
          />
        </a>
        <a
          className="socialCircle"
        >
          <img
            alt="twitter"
            src="https://upload.wikimedia.org/wikipedia/commons/thumb/4/4f/Twitter-logo.svg/2491px-Twitter-logo.svg.png"
          />
        </a>
        <a
          className="socialCircle"
        >
          <img
            alt="instagram"
            src="https://upload.wikimedia.org/wikipedia/commons/thumb/e/e7/Instagram_logo_2016.svg/2048px-Instagram_logo_2016.svg.png"
          />
        </a>
        <a
          className="socialCircle"
        >
          <img
            alt="website"
            src="https://w7.pngwing.com/pngs/549/715/png-transparent-web-development-logo-website-web-design-symmetry-internet.png"
          />
        </a>
      </div>
    </div>
  </div>
  <div
    className="body"
  >
    <div
      className="description"
    >
      <p />
      <p>
        <strong>
          Category:
        </strong>
         
      </p>
      <p>
        <strong>
          Followers:
        </strong>
         
      </p>
    </div>
    <div
      className="events"
    >
      <h2>
        Events:
      </h2>
      <div
        className="events"
        data-testid="events-list"
      />
    </div>
  </div>
  <div
    style={
      Object {
        "backgroundColor": "#abc4ff",
        "borderColor": "gray",
        "borderRadius": 20,
        "borderWidth": 2,
      }
    }
  >
    <h3>
      <center>
        Contact: 
      </center>
    </h3>
    <form
      onSubmit={[Function]}
      style={
        Object {
          "marginLeft": 38,
          "width": "95%",
        }
      }
    >
      <div
        className="form-group"
      >
        <label
          htmlFor="customerName"
        >
          Your Name
        </label>
        <input
          className="form-control"
          name="customerName"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="customerEmail"
        >
          Your Email
        </label>
        <input
          className="form-control"
          name="customerEmail"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="messageSubject"
        >
          Subject of your query
        </label>
        <input
          className="form-control"
          name="messageSubject"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="message"
        >
          Your query
        </label>
        <textarea
          className="form-control"
          cols="173"
          id="message"
          name="message"
          onBlur={[Function]}
          onChange={[Function]}
          rows="8"
          style={
            Object {
              "marginBottom": "8px",
            }
          }
          type="text"
          value=""
        />
      </div>
      <button
        className="button"
        style={
          Object {
            "marginBottom": "8px",
          }
        }
      >
        Submit query
      </button>
    </form>
  </div>
  <a
    href="/edit-society/undefined"
    onClick={[Function]}
  >
    <button
      className="button"
    >
      Edit Society
    </button>
  </a>
</div>
`;
